File: go-backend-interview-prep/08_http_and_backend/03_middleware_basics/main.go
package main

import (
	"fmt"
	"log"
	"net"
	"net/http"
	"time"
)

// --- Middleware: logging ---
func loggingMW(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		next.ServeHTTP(w, r)
		fmt.Printf("  [LOG] %s %s  %s\n", r.Method, r.URL.Path, time.Since(start))
	})
}

// --- Middleware: timing header ---
func timingMW(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		next.ServeHTTP(w, r)
		w.Header().Set("X-Response-Time", time.Since(start).String())
	})
}

// --- Middleware: panic recovery ---
func recoveryMW(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer func() {
			if err := recover(); err != nil {
				fmt.Printf("  [RECOVER] panic: %v\n", err)
				http.Error(w, `{"error":"internal server error"}`, http.StatusInternalServerError)
			}
		}()
		next.ServeHTTP(w, r)
	})
}

// chain applies middleware in order: first listed = outermost
func chain(h http.Handler, mws ...func(http.Handler) http.Handler) http.Handler {
	for i := len(mws) - 1; i >= 0; i-- {
		h = mws[i](h)
	}
	return h
}

// --- Handlers ---

func helloHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	fmt.Fprintln(w, "Hello from middleware demo!")
}

func panicHandler(w http.ResponseWriter, r *http.Request) {
	panic("something went wrong!")
}

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/hello", helloHandler)
	mux.HandleFunc("/panic", panicHandler)

	// Chain: recovery -> logging -> timing -> mux
	handler := chain(mux, recoveryMW, loggingMW, timingMW)

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()
	fmt.Printf("Server on http://%s\n", addr)
	fmt.Println("  GET /hello  -> greeting (logged + timed)")
	fmt.Println("  GET /panic  -> triggers recovery middleware")

	server := &http.Server{Handler: handler}
	go func() {
		time.Sleep(30 * time.Second)
		fmt.Println("\nAuto-shutdown (demo)...")
		server.Close()
	}()

	// Demo: self-test
	go func() {
		time.Sleep(100 * time.Millisecond)
		base := "http://" + addr
		fmt.Println("\n--- Demo requests ---")

		resp, _ := http.Get(base + "/hello")
		if resp != nil {
			resp.Body.Close()
			fmt.Println("  /hello -> status", resp.StatusCode)
		}

		resp2, _ := http.Get(base + "/panic")
		if resp2 != nil {
			resp2.Body.Close()
			fmt.Println("  /panic -> status", resp2.StatusCode, "(recovered)")
		}

		fmt.Println("\nDemo done. Server still running (30s timeout)...")
	}()

	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}

File: go-backend-interview-prep/08_http_and_backend/03_middleware_basics/main.py
"""Middleware basics -- Python equivalent of the Go example.

Python stdlib has no middleware chain. We simulate it by wrapping
handler methods with decorators / manual composition.
"""

import json
import threading
import time
import traceback
import urllib.request
from http.server import HTTPServer, BaseHTTPRequestHandler


# --- Middleware-style wrappers ---
# In Python stdlib, middleware = wrapping handler logic manually.

class MyHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        # --- Recovery middleware (try/except around everything) ---
        try:
            start = time.perf_counter()
            self._route()
            elapsed = time.perf_counter() - start
            # --- Logging middleware ---
            print(f"  [LOG] {self.command} {self.path}  {elapsed*1000:.1f}ms")
        except Exception:
            print(f"  [RECOVER] panic: {traceback.format_exc().splitlines()[-1]}")
            self.send_response(500)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"error": "internal server error"}).encode())

    def _route(self):
        if self.path == "/hello":
            self.send_response(200)
            self.send_header("Content-Type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Hello from middleware demo!\n")

        elif self.path == "/panic":
            raise RuntimeError("something went wrong!")

        else:
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b"not found\n")

    def log_message(self, *args):
        pass


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), MyHandler)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"Server on {base}")
    print("  GET /hello  -> greeting (logged + timed)")
    print("  GET /panic  -> triggers recovery middleware")
    threading.Thread(target=server.serve_forever, daemon=True).start()

    # --- Demo requests ---
    print("\n--- Demo requests ---")

    resp = urllib.request.urlopen(f"{base}/hello", timeout=5)
    print(f"  /hello -> status {resp.status}")

    try:
        urllib.request.urlopen(f"{base}/panic", timeout=5)
    except urllib.error.HTTPError as e:
        print(f"  /panic -> status {e.code} (recovered)")

    server.shutdown()
    print("\nDone.")


if __name__ == "__main__":
    main()
